File: src/com/kopiright/vkopi/lib/ui/vaadin/visual/VWindowController.java
/*
 * Copyright (c) 1990-2009 kopiRight Managed Solutions GmbH
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * $Id$
 */

package com.kopiright.vkopi.lib.ui.vaadin.visual;

import org.kopi.vaadin.addons.PopupWindow;

import com.kopiright.vkopi.lib.preview.VPreviewWindow;
import com.kopiright.vkopi.lib.ui.vaadin.base.BackgroundThreadHandler;
import com.kopiright.vkopi.lib.visual.ApplicationContext;
import com.kopiright.vkopi.lib.visual.VException;
import com.kopiright.vkopi.lib.visual.VHelpViewer;
import com.kopiright.vkopi.lib.visual.VMenuTree;
import com.kopiright.vkopi.lib.visual.VRuntimeException;
import com.kopiright.vkopi.lib.visual.VWindow;
import com.kopiright.vkopi.lib.visual.WindowBuilder;
import com.kopiright.vkopi.lib.visual.WindowController;

/**
 * The <code>VWindowController</code> is the vaadin implementation
 * of the {@link WindowController} specifications.
 */
@SuppressWarnings("serial")
public class VWindowController extends WindowController {

  //---------------------------------------------------
  // IMPLEMENTATIONS
  //---------------------------------------------------
  
  @Override
  public boolean doModal(VWindow model) {
    ModalViewRunner		viewStarter;
    
    viewStarter = new ModalViewRunner(model);
    BackgroundThreadHandler.startAndWait(viewStarter, model);
    
    return (viewStarter.getView() == null) ? false : viewStarter.getView().getReturnCode() == VWindow.CDE_VALIDATE;
  }

  @Override
  public void doNotModal(VWindow model) {
    WindowBuilder		builder;
    
    builder = getWindowBuilder(model);
    if (builder != null) {
      try {
	DWindow			view;
	VApplication		application;

	view = (DWindow) builder.createWindow(model);
	view.run();
	application = (VApplication) ApplicationContext.getApplicationContext().getApplication();
	if (application != null) {
	  if (model instanceof VPreviewWindow
	      || model instanceof VHelpViewer
	      || model instanceof VMenuTree)
	  {
	    showNotModalPopupWindow(application, view, model.getTitle());
	  } else {
	    application.addWindow(view);
	  }
	}
      } catch (VException e) {
	throw new VRuntimeException(e.getMessage(), e);
      }
    }
  }
  
  /**
   * Shows a modal window in a popup view. This will handle
   * a window view not in a tabsheet but in a non modal
   * popup window.
   * @param application The application instance.
   * @param view The window view.
   * @param title The window title.
   */
  protected void showNotModalPopupWindow(final VApplication application,
                                         DWindow view,
                                         String title)
  {
    if (application == null) {
      return;
    }
    
    final PopupWindow			popup;
	  
    popup = new PopupWindow();
    popup.setModal(false);
    popup.setContent(view);
    popup.setCaption(title); // put popup title
    BackgroundThreadHandler.access(new Runnable() {
      
      @Override
      public void run() {
	application.attachComponent(popup);
	application.push(); // a push is needed to see the popup.
      }
    });
  }

  //---------------------------------------------------
  // MODAL VIEW STARTER
  //---------------------------------------------------

  /**
   * A modal view runner background task.
   */
  /*package*/ class ModalViewRunner implements Runnable {

    //---------------------------------------
    // CONSTRUCTOR
    //---------------------------------------
    
    /**
     * Creates a new <code>ModalViewRunner</code> runner.
     * @param model The window model.
     */
    public ModalViewRunner(VWindow model) {
      this.model = model;
    }

    //---------------------------------------
    // IMPLEMENTATION
    //---------------------------------------
    
    @Override
    public void run() {
      WindowBuilder		builder;
      
      builder = getWindowBuilder(model);
      if (builder != null) {
	
	try {
	  VApplication		application;
	      
	  view = (DWindow) builder.createWindow(model);
	  view.run();
	  application = (VApplication) ApplicationContext.getApplicationContext().getApplication();
	  if (application != null) {
	    PopupWindow		popup;
		  
	    popup = new PopupWindow();
	    popup.setModal(true);
	    popup.setContent(view);
	    popup.setCaption(model.getTitle()); // put popup title
	    application.attachComponent(popup);
	    application.push(); // a push is needed to see the popup.
	  }
	} catch (VException e) {
	  throw new VRuntimeException(e.getMessage(), e);
	}
      }
    }

    /**
     * Returns the window view.
     * @return The window view.
     */
    public DWindow getView() {
      return view;
    }

    //---------------------------------------
    // DATA MEMBERS
    //---------------------------------------

    private DWindow       		view;
    private final VWindow       	model; 
  }
}

File: src/org/kopi/vkopi/lib/ui/vaadinflow/visual/VWindowController.kt
/*
 * Copyright (c) 2013-2021 kopiLeft Services SARL, Tunis TN
 * Copyright (c) 1990-2021 kopiRight Managed Solutions GmbH, Wien AT
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.kopi.vkopi.lib.ui.vaadinflow.visual

import org.kopi.vkopi.lib.form.VForm
import org.kopi.vkopi.lib.preview.VPreviewWindow
import org.kopi.vkopi.lib.ui.vaadinflow.base.BackgroundThreadHandler
import org.kopi.vkopi.lib.ui.vaadinflow.base.BackgroundThreadHandler.accessAndAwait
import org.kopi.vkopi.lib.ui.vaadinflow.base.BackgroundThreadHandler.accessAndPush
import org.kopi.vkopi.lib.ui.vaadinflow.base.BackgroundThreadHandler.startAndWaitAndPush
import org.kopi.vkopi.lib.ui.vaadinflow.field.TextField
import org.kopi.vkopi.lib.ui.vaadinflow.form.DField
import org.kopi.vkopi.lib.ui.vaadinflow.form.DGridEditorField
import org.kopi.vkopi.lib.ui.vaadinflow.window.PopupWindow
import org.kopi.vkopi.lib.visual.VException
import org.kopi.vkopi.lib.visual.VHelpViewer
import org.kopi.vkopi.lib.visual.VMenuTree
import org.kopi.vkopi.lib.visual.VRuntimeException
import org.kopi.vkopi.lib.visual.VWindow
import org.kopi.vkopi.lib.visual.WindowController

import com.vaadin.flow.component.Component

/**
 * The `VWindowController` is the vaadin implementation
 * of the [WindowController] specifications.
 */
class VWindowController : WindowController() {
  //---------------------------------------------------
  // IMPLEMENTATIONS
  //---------------------------------------------------
  override fun doModal(model: VWindow): Boolean {
    return try {
      val viewStarter = ModalViewRunner(model)
      startAndWaitAndPush(model as Object) {
        viewStarter.run()
      }
      if (viewStarter.getView() == null) false else viewStarter.getView()!!.returnCode == VWindow.CDE_VALIDATE
    } finally {
      // This is a turn around to kill delayed wait dialog displayed in modal windows
      model.unsetWaitInfo()
    }
  }

  override fun doNotModal(model: VWindow) {
    val builder = getWindowBuilder(model)
    if (builder != null) {
      try {
        lateinit var view: DWindow

        accessAndAwait {
          view = builder.createWindow(model) as DWindow
        }

        view.run()
        val application = getApplication()
        if (application != null) {
          if (model is VPreviewWindow
            || model is VHelpViewer
            || model is VMenuTree
          ) {
            showNotModalPopupWindow(view, model.getTitle())
          } else {
            application.addWindow(view, model.getTitle())
          }
        }
      } catch (e: VException) {
        throw VRuntimeException(e.message, e)
      }
    }
  }

  /**
   * Shows a modal window in a popup view. This will handle
   * a window view not in a tabsheet but in a non modal
   * popup window.
   * @param application The application instance.
   * @param view The window view.
   * @param title The window title.
   */
  protected fun showNotModalPopupWindow(
          view: DWindow,
          title: String,
  ) {
    val popup = PopupWindow(getApplication()?.mainWindow)
    accessAndPush {
      popup.isModal = false
      popup.setContent(view)
      popup.setCaption(title) // put popup title
      popup.open()

      // Focus on a field inside a popup is not working.
      // Call focus asynchronously after some time from popup attachment.
      view.focusOnActiveField()
    }
  }

  private fun getApplication(): VApplication? {
    val ui = BackgroundThreadHandler.locateUI()
    return if (ui == null) {
      null
    } else {
      ui.children
        .filter { component -> component is VApplication }
        .findFirst()
        .orElse(null) as? VApplication
    }
  }

  //---------------------------------------------------
  // MODAL VIEW STARTER
  //---------------------------------------------------
  /**
   * A modal view runner background task.
   */
  internal inner class ModalViewRunner(val model: VWindow) : Runnable {

    private var view: DWindow? = null

    //---------------------------------------
    // IMPLEMENTATION
    //---------------------------------------
    override fun run() {
      val builder = getWindowBuilder(model)
      if (builder != null) {
        try {
          view = builder.createWindow(model) as DWindow
          view!!.run()
          val application = getApplication()
          if (application != null) {
            val popup = PopupWindow(application.mainWindow)
            popup.isModal = true
            popup.setContent(view!!)
            popup.setCaption(model.getTitle()) // put popup title
            popup.open()

            // Focus on a field inside a popup is not working.
            // Call focus asynchronously after some time from popup attachment.
            view!!.focusOnActiveField()
          }
        } catch (e: VException) {
          throw VRuntimeException(e.message, e)
        }
      }
    }

    /**
     * Returns the window view.
     * @return The window view.
     */
    fun getView(): DWindow? {
      return view
    }
  }

  /**
   * Ensure that active field of this window is focused.
   *
   * This is a workaround for issue https://github.com/vaadin/vaadin-dialog-flow/issues/55.
   */
  fun DWindow.focusOnActiveField() {
    element.executeJs("").then {
      val lasFocusedField = (getModel() as VForm).getActiveBlock()?.activeField?.getDisplay()

      if (lasFocusedField != null) {
        val internalField: Component = when (lasFocusedField) {
          is DField -> (lasFocusedField.wrappedField as TextField).inputField
          is DGridEditorField<*> -> lasFocusedField.editor
          else -> return@then
        }
        internalField.element.executeJs("setTimeout(function(){$0.focus()},60)", internalField.element)
      }
    }
  }
}
